helpers/embeds.js

```diff
@@ -0,0 +1,78 @@
+import { findComposedAncestor } from '../helpers/dom.js';
+import { html } from 'lit';
+import { tryGetPluginByKey } from '@brightspace-ui/core/helpers/plugins.js';
+
+class RenderNode {
+	#childNodes;
+	#node;
+
+	constructor(node) {
+		this.#node = node;
+		this.#childNodes = [];
+	}
+
+	addChildNode(childNode) {
+		this.#childNodes.push(childNode);
+	}
+
+	async renderNode(options) {
+		for await (const childNode of this.#childNodes) {
+			await childNode.renderNode(options);
+		}
+
+		if (this.#node.id === '3') {
+			console.log('wee');
+		}
+
+		const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', this.#node.dataset.d2lEmbedType);
+		if (!embedRendererPlugin) return;
+
+		const fragment = new DocumentFragment();
+		this.#node.childNodes.forEach(node => {
+			const temp = node.cloneNode(true);
+			fragment.appendChild(temp);
+		});
+
+		let props = this.#node.dataset.d2lEmbedData;
+		try {
+			props = JSON.parse(this.#node.dataset.d2lEmbedData);
+		} catch (e) { /* empty */ }
+
+		const embed = await embedRendererPlugin.renderView(fragment, props || {}, options || {});
+		if (embed) this.#node.replaceWith(embed);
+	}
+
+}
+
+export async function createEmbedPlaceholder(embedType, options) {
+	const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', embedType);
+	if (!embedRendererPlugin) return;
+
+	const placeholderData = await embedRendererPlugin.getPlaceholder(options);
+	/* NOT IMPLEMENTED */
+}
+
+export async function renderEmbeds(node, options) {
+	const elems = [...node.querySelectorAll('div[data-d2l-embed-type], span[data-d2l-embed-type]')];
```

Comment:
I can get rid of the array construction here. It was present because I was using other array functions that aren't also present on `NodeList`. Now I'm not. Since this list will always be static, no need to massage it into an array.

Reply:
Oop, scratch that. Forgot that I'm still using `includes`.

Reply:
Yeah, I'm a bit surprised that more array functions aren't implemented on `NodeList`. Maybe `includes` is a bit of a weird one, but no `map`!? Sad.

Reply:
I'm not opposed to querying `div` and `span` like this, but we could perhaps just query `[data-d2l-embed-type]`. Is there a reason to limit it?

Reply:
So the idea (not currently implemented in this PR) is that we have a helper function that builds the div/span for them. That gives an option for the placeholder to be either inline or not, allows us to make some assumption when building those elements for the consumer, and also makes it clear that the placeholder shouldn't really be doing anything super complex.